Multiple categories and category related changes

okkez opened the pull request:
connect to https://github.com/bm-sms/daimon-news-misc/issues/154
- multiple categories
- design for category related changes
  - category hierarchy
  - multiple categories

okkez committed:
Select multiple categories

okkez committed:
Follow previous changes related to Groonga

okkez committed:
Follow previous changes related to view

okkez committed:
Update seed to use multiple categories

okkez committed:
Fix validation

okkez committed:
Setup order value according to target

okkez committed:
Add scope to fetch categorized posts

okkez committed:
Use categories

okkez committed:
Use categorized scope

But this action does not work well...

okkez committed:
Use categories

okkez committed:
Use `.search-result-list`

Because we have multiple categories under another `<ul>`.

okkez committed:
Fix integration test for multiple categories

okkez committed:
Simplify CurrentCategoryTest

We don't need register records via management console.

okkez committed:
Add parentheses

okkez committed:
Add parentheses

okkez committed:
Use categories

okkez committed:
Use `<ul>` to display categories on `/editor/posts`

okkez committed:
Set `list-style-type: none` in `/editor/posts`

okkez committed:
Set `categorization.order` properly

okkez committed:
Apply toArray for jQuery object

In previous version, `Array.from(orders)` does not work on phantomjs.
Because phantomjs does not support ES2015(ES6) features.
So I convert jQuery object to Array before apply _toConsumableArray.

okkez committed:
Set categories via categorizations

okkez committed:
Remove needless code

Follow previous changes.

okkez committed:
Display related categories after content

okkez committed:
Fix scope Post.categorized

In this version, we can fetch categorized posts that have multiple
categories.

okkez committed:
Add seed

Posts have multiple categories.

okkez committed:
Remove category description from category list

Because it is redundant.

okkez committed:
Fix a typo

okkez committed:
Arrange design

okkez committed:
Display recent posts in `categories#show`

okkez committed:
Display related categories before related posts

okkez committed:
Extract partial template for related categories

okkez committed:
Use ordered list (ol)

okkez committed:
Do not display categories that have no published posts

Display category when the category's children have published posts.

okkez committed:
Display root categories properly

In previous version, display category that has no posts.

okkez committed:
Fix `Category#post_count`

Count posts in all subtree.

okkez committed:
Display post that does not have categories properly

okkez committed:
Use pre-loaded record object

okkez committed:
Simplify

okkez committed:
Remove unnecessary heading

okkez committed:
Cosmetic changes

okkez committed:
Rename scope

categorized ->
categorized_by

okkez committed:
Use `lambda` instead of `->`

This suppresses rubocop warnings such as:

* Style/Lambda: Use the lambda method for multiline lambdas.
* Style/IndentationWidth: Use 2 (not -13) spaces for indentation.

okkez commented:
Depends on #582

okkez committed:
Extract migration for #582

We should merge this commit before merge #582

okkez committed:
Add models related to multiple categories

okkez committed:
Remove null constraint from `posts.category_id`

Because we will change seed data in #582.

okkez committed:
Remove microdata from related categories

tricknotes commented:
![2016-08-02 13 50 43](https://cloud.githubusercontent.com/assets/290782/17317401/42661d96-58b8-11e6-8921-d5d14ba87542.png)

tricknotes commented:
In child category, the parent category in menu bar should be highlighted?
![2016-08-02 14 12 39](https://cloud.githubusercontent.com/assets/290782/17317690/63f15392-58bb-11e6-9051-4ccd1cd213b3.png)

tricknotes commented on app/assets/javascripts/editor/form.js.es6:
```diff
@@ -25,6 +25,7 @@ $(() => {
 
     applySelect2($element.find('select.select2'));
-    setupCreditOrder($element);
+    setupOrder($element, '.credit-order-value');
```
Can you use purpose specific class name here?

okkez replied:
What did you mean?
I want isolated scope for `orders` in `setUpOrder`.

tricknotes replied:
Ah, right. Sorry for my unclear comment :<

This method name is pretty good!
Can you extract the abstract class name that means`'.credit-order-value` and `'.categorization-order-value'`?

tricknotes replied:
An, I've a misunderstanding :<
This way is no any problem!

tricknotes commented on app/models/category.rb:
```diff
@@ -46,14 +47,19 @@ def full_name
-      # Use `to_a` to call `Enumerable#sum` here.
-      # We can remove `to_a` after Rails 5. http://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-sum
-      children.to_a.sum {|child| child.posts.published.count }
+      ids = subtree.inject([]) do |memo, category|
```
`AR::Relation#select` returns AR object. However it is too large object to be needed.

How about the following?

``` ruby
subtree.flat_map {|category| category.posts.published.pluck(:id) }.uniq.count
```

okkez replied:
Ah, good catch.

tricknotes commented:
`post.categories.first` appears in many place.
This is not just first, it means main category.

Can you extract it as the following method?
``` ruby
class Post
  def main_category
    categories.first!
  end
end
```

tricknotes commented on app/decorators/post_decorator.rb:
```diff
@@ -22,7 +22,7 @@ def to_og_params
   def to_article_params
     {
-      section: category.name,
+      section: categories.first&.name,
```
Is this lonely operator is required?

okkez replied:
For now, we need this.
Because we can register posts without categories.

tricknotes replied:
Category is required by https://github.com/bm-sms/daimon-news/pull/599

tricknotes commented on app/groonga/post_indexer.rb:
```diff
@@ -11,11 +11,8 @@ def initialize
-    else
-      category = nil
+    categories = post.categories.map do |category|
+      @categories.add(category.id, name: category.name)
```
`map` with side effect is bad smell :<

tricknotes commented on app/models/category.rb:
```diff
@@ -46,14 +47,19 @@ def full_name
     end
   end
 
+  def children_or_posts?
```
`has_children_or_posts?` ?

okkez replied:
Rubocop says 

> C: Style/PredicateName: Rename has_children_or_posts? to children_or_posts?.

tricknotes replied:
Ouch!

tricknotes commented on app/models/post.rb:
```diff
@@ -23,8 +23,13 @@ def with_ordered
   scope :published, -> { where("published_at <= ?", Time.current) }
   scope :order_by_recent, -> { order(published_at: :desc, id: :asc) }
+  scope :categorized_by, lambda {|category|
+    category_ids = category.has_children? ? category.subtree_ids : category.id
```
If this variable name is `category_ids`, can you always assign an instance of `Array`?

``` diff
- category.id
+ [category.id]
```

tricknotes commented on app/views/editor/posts/_form.html.haml:
```diff
@@ -23,7 +23,13 @@
         = link_to_add_association 'Add credit', f, :credits, class: 'btn btn-default pull-right'
 
-  = f.collection_select :category_id, current_site.categories.leaves, :id, :full_name, {}, {class: "select2", "data-select2-allow-clear": false}
+  = f.form_group label: {text: :categorizations}, help: "カテゴリを追加します" do
```
- There are no need to use Symbol for text.
- This text should be a "Categories", not a "Categorizations".

``` diff
- {text: :categorizations}
+ {text: 'Categories'}
```

okkez commented:
> In child category, the parent category in menu bar should be highlighted?

Good catch. I will fix this later.
But PRs are welcome!

tricknotes commented:
The sub categories style is a bit strange :<

![2016-08-02 15 36 24](https://cloud.githubusercontent.com/assets/290782/17319198/eaf9fe10-58c6-11e6-936b-d475ae544ccf.png)

okkez committed:
Use `pluck(:id)` instead of `select(:id)`

Because `select(:id)` returns AR object.

okkez committed:
Assume that `category_ids` is an `Array`

okkez committed:
Use String instead of Symbol

okkez committed:
Highlight root category of current category on menu bar

okkez committed:
Use `#each` instead of `#map`

kou commented on app/groonga/post_indexer.rb:
```diff
@@ -11,23 +11,22 @@ def initialize
-                        name: credit.participant.name,
-                        summary: credit.participant.summary)
+    participants = []
+    post.credits.map do |credit|
```
`#map` is still used.

tricknotes committed:
Add validation for `Post#categorizations` to have one or more item

tricknotes committed:
Update seed to create post with category in atomically

tricknotes commented on app/helpers/current_resource_helper.rb:
```diff
@@ -10,6 +10,6 @@ def current_site
 
   def current_category
-    nil
+    @current_category
```
Can you overwrite in each controller?

---

共通モジュール内のインスタンス変数に値を代入するよりも、必要に応じて各コントローラでメソッドをオーバライドする方がモジュールの API としては整理されたものになるかと思いますー

okkez committed:
Set current category properly

okkez committed:
Add `CategoryDecorator#highlighted?`

We cannot assume that current_category is present.
Because `CurrentResourceHelper#current_category` returns `nil`.

tricknotes committed:
Fix test to create categorization with post

tricknotes committed:
Show error message for categorization

okkez committed:
Fix custom validation

In previous version, we can register post with category has children.

okkez committed:
Add validation

In previous version, raise error when create/update post that has same
category twice.

okkez committed:
Use meaningful name

okkez committed:
Use Post custom validation instead of Categorization's validation

Because in previous version, we cannot validate multiple same categories
when create/update Post.

okkez committed:
Category is required now

See #599

okkez committed:
Add `Post#main_category` and use it

okkez committed:
Remove useless assignment

okkez committed:
Suppress rubocop warning Style/MultilineMethodCallBraceLayout

okkez committed:
Use feature flag for multiple categories

tricknotes committed:
Fix markup to keep backward compatibility

tricknotes committed:
Add sub header for posts

tricknotes committed:
[Refactor] Move header to partial

tricknotes committed:
[Refactor] `next` in view is difficult to me

tricknotes committed:
[Refactor] Make partial inline

Partial should be a private view in controller scope.

tricknotes committed:
[Refactor] Use class name to apply style

okkez committed:
Stop displaying duplicated posts on `categories#show` post list when it has multiple categories

In previous version, duplicated posts are displayed.

tricknotes committed:
Specify class name to independ content type

This list may have post or category.
However style is shared these items.

okkez committed:
Add feature flag for hierarchical categories

tricknotes committed:
Fix sub category's text :<

okkez committed:
Revert "Use `to_sql` to reduce SQL"

This reverts commit 6546a6aadf8fb3cbe05da323cb436849c7906d82.

Because it does not work.

Generated SQL as following:

```
SELECT
 "categories"."id" AS t0_r0,
 "categories"."name" AS t0_r1,
 "categories"."description" AS t0_r2,
 "categories"."created_at" AS t0_r3,
 "categories"."updated_at" AS t0_r4,
 "categories"."site_id" AS t0_r5,
 "categories"."slug" AS t0_r6,
 "categories"."order" AS t0_r7,
 "categories"."ancestry" AS t0_r8,
 "posts"."id" AS t1_r0,
 "posts"."title" AS t1_r1,
 "posts"."body" AS t1_r2,
 "posts"."site_id" AS t1_r3,
 "posts"."created_at" AS t1_r4,
 "posts"."updated_at" AS t1_r5,
 "posts"."published_at" AS t1_r6,
 "posts"."category_id" AS t1_r7,
 "posts"."thumbnail" AS t1_r8,
 "posts"."original_updated_at" AS t1_r9,
 "posts"."public_id" AS t1_r10,
 "posts"."serial_id" AS t1_r11
FROM
 "categories"
 LEFT OUTER JOIN
 "categorizations"
   ON "categorizations"."category_id" = "categories"."id"
 LEFT OUTER JOIN
 "posts"
   ON "posts"."id" = "categorizations"."post_id"
WHERE
 "posts"."id" IS NULL
 AND ("categories"."id" != $1)
 ORDER BY "categories".ancestry NULLS FIRST, "categories"."order" ASC  [["id", 0]]
```

tricknotes committed:
Add more link

okkez committed:
Use feature flag hierarchical_categories_enabled

tricknotes committed:
Revert "Use feature flag for multiple categories"

This reverts commit b9444fcb781d3cde5faeee4f52e9ca6597c676b0.
This feature doesn't break anything.

mtsmfm commented on app/models/category.rb:
```diff
@@ -40,20 +41,30 @@ def siblings
+      ids = subtree.flat_map do |category|
+        category.posts.published.pluck(:id)
+      end
+      ids.uniq.size
```
`Post.joins(:categories).merge(subtree).published.uniq.count` ?

tricknotes committed:
Specify `Site#hierarchical_categories_enabled` flag to development seed

tricknotes committed:
Add class to apply style

tricknotes committed:
[Refactor] Remove dup of DOM structure

tricknotes committed:
Fix current link in related categories' breadcrumbs

tricknotes committed:
[Refactor] Don't pass extra option to gretel

tricknotes committed:
Update microdata using http://schema.org

tricknotes committed:
Add missing tailing semicolon.

tricknotes committed:
[Refactor] :golf:

tricknotes committed:
[Refactor] `uniq` should be called with `joins` in this case

tricknotes committed:
Update comment

We should remove this method when a new ancestry is released.

mtsmfm committed:
Cosme DOM structure

tricknotes committed:
[Refactor] Don't use view specific method name

tricknotes committed:
[Refactor] Remove unused local variables

mtsmfm committed:
Merge pull request #607 from bm-sms/update-microdata

Update microdata

tricknotes committed:
Update daimon_news-layout gem

To use https://github.com/bm-sms/daimon_news-layout/pull/24

tricknotes commented:
🚢
